shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java

```diff
@@ -0,0 +1,141 @@
+package org.metaborg.spoofax.shell.client;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import org.metaborg.core.completion.ICompletionService;
+import org.metaborg.spoofax.core.unit.ISpoofaxParseUnit;
+
+import jline.console.ConsoleReader;
+
+/**
+ * A terminal UI which is both an {@link IEditor} and {@link IDisplay}.
+ */
+public class TerminalUserInterface implements IEditor, IDisplay {
+    private ConsoleReader reader;
+    private String prompt;
+    private String continuationPrompt;
+    private ArrayList<String> lines;
+    private PrintWriter out;
+    private PrintWriter err;
+
+    /**
+     * @throws IOException
+     *             when an IO error occurs.
+     */
+    public TerminalUserInterface() throws IOException {
+        this(System.in, System.out, System.err);
+    }
+
+    /**
+     * @param in
+     *            The {@link InputStream} from which to read user input
+     * @param out
+     *            The {@link PrintStream} to write results to.
+     * @param err
+     *            The {@link PrintStream} to write errors to.
+     * @throws IOException
+     *             when an IO error occurs.
+     */
+    public TerminalUserInterface(InputStream in, OutputStream out, OutputStream err)
+        throws IOException {
+        reader = new ConsoleReader(in, out);
+        reader.setExpandEvents(false);
+        reader.setHandleUserInterrupt(true);
+        reader.setBellEnabled(true);
+        this.out =
+            new PrintWriter(new BufferedWriter(new OutputStreamWriter(out,
+                                                                      Charset.forName("UTF-8"))));
+        this.err =
+            new PrintWriter(new BufferedWriter(new OutputStreamWriter(err,
+                                                                      Charset.forName("UTF-8"))));
+        setPrompt(">>> ");
+        setContinuationPrompt("... ");
+        lines = new ArrayList<>();
+    }
+
+    /**
+     * Save this line as the end of the multi-line input.
+     *
+     * @param lastLine
+     *            the line to save.
+     */
+    protected void saveLine(String lastLine) {
+        lines.add(lastLine);
+    }
+
+    /**
+     * Set the prompt to display.
+     *
+     * @param promptString
+     *            The prompt string.
+     */
+    public void setPrompt(String promptString) {
+        prompt = promptString;
```

Comment:
Why not directly use `reader.setPrompt`? Also, can the reader handle a `null` prompt?